### PR TITLE
136 implement course view

### PR DESCRIPTION
**Pull request by Sainaamr**
**What was done:**

- create a UI for the detail course
- fetching the streams for a course and updating the view accordingly 
- fetching the thumbnails live or not 
- showing the thumbnails

**things that most likely going to change:** 

- how the thumbnails look (waiting to see the course over view thumbnails and then change accordingly)

### CODE REVIEW

**Comment by cbetul** [comment]
I think it looks pretty good, I haven't come across any functional flaws or bugs. I just have a few suggestions:
- The navigation bar should look like in the Figma (search input is not hidden, back arrow icon etc.), for example, it should look like this:
Screenshot 2023-12-27 at 21 13 10

- You can adjust the padding of the course cards, it looked like slightly overloaded horizontally to me.

File: lib/models/user/user_state_model.dart [hunk 1]
```diff
@@ -16,6 +16,9 @@ class UserState {
   final AppError? error;
   final List<Course>? downloadedCourses;
 
+  //final List<Stream>? courseStreams; // New property for course streams
```
Comment (cbetul) [inline lib/models/user/user_state_model.dart:19]: don't forget to remove comments :)

File: lib/views/course_detail_view/stream_card.dart [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                  return Image.asset(
+                    'assets/images/default_image.png',
+                    // Fallback image in case of error
+                    fit: BoxFit.cover,
+                  );
+                },
+              ),
+            ),
+            const Padding(
+              padding: EdgeInsets.fromLTRB(8.0, 8.0, 8.0, 12.0),
```
Comment (cbetul) [inline lib/views/course_detail_view/stream_card.dart:74]: You can adjust the padding of the course cards, it looked like slightly overloaded horizontally to me.

File: lib/views/course_detail_view/detail_course_List.dart [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    this.onRefresh,
+  });
+
+  @override
+  Widget build(BuildContext context, WidgetRef ref) {
+    return BaseView(
+      title: title,
+      actions: [
+        IconButton(
+          icon: const Icon(Icons.search),
```
Comment (cbetul) [inline lib/views/course_detail_view/detail_course_List.dart:24]: I think it looks pretty good, I haven't come across any functional flaws or bugs. I just have a few suggestions:
- The navigation bar should look like in the Figma (search input is not hidden, back arrow icon etc.), for example, it should look like this:
Screenshot 2023-12-27 at 21 13 10

**Review by cbetul — CHANGES_REQUESTED** [review]
I think it looks pretty good, I haven't come across any functional flaws or bugs. I just have a few suggestions:

**Review by Anishyou — CHANGES_REQUESTED** [review]
There is a bug where I cannot go back when I click on a course. THere is no back button, no nav bar nothing?






https://github.com/TUM-Dev/gocast_mobile/assets/123313052/14e32cea-8f1c-4ec2-8598-64b433eceee4

**Comment by Sainaamr** [comment]
> There is a bug where I cannot go back when I click on a course. THere is no back button, no nav bar nothing?
> 
>  Simulator.Screen.Recording.-.iPhone.15.Pro.Max.-.2023-12-27.at.21.41.54.mp4


https://github.com/TUM-Dev/gocast_mobile/assets/90347142/d5de32dd-6c64-48da-bbc9-070e7605216c

I dont have the bug. i ran it on iphone 15 pro max too.

**Comment by Anishyou** [comment]
> > There is a bug where I cannot go back when I click on a course. THere is no back button, no nav bar nothing?
> > Simulator.Screen.Recording.-.iPhone.15.Pro.Max.-.2023-12-27.at.21.41.54.mp4
> 
>  Screen.Recording.2023-12-28.at.12.08.46.mov 
> I dont have the bug. i ran it on iphone 15 pro max too. do you know where is the problem?

Firstly I think you should update your branch with dev. I have solved the navigation issues.

**Comment by Anishyou** [comment]
> > There is a bug where I cannot go back when I click on a course. THere is no back button, no nav bar nothing?
> > Simulator.Screen.Recording.-.iPhone.15.Pro.Max.-.2023-12-27.at.21.41.54.mp4
> 
>  Screen.Recording.2023-12-28.at.12.08.46.mov 
> I dont have the bug. i ran it on iphone 15 pro max too. do you know where is the problem?

I checked again. As I sent the video, error is still there. which branch are you working on?
136-implement-course-view right?

**Comment by Anishyou** [comment]
Screenshot 2023-12-28 at 13 38 02

I would suggest to add some mock data for course video so that we see how it looks. It is not needed now , you can create another PR for it also.

**Comment by Sainaamr** [comment]
> Screenshot 2023-12-28 at 13 38 02
> I would suggest to add some mock data for course video so that we see how it looks. It is not needed now , you can create another PR for it also.

i have a question why dont u have any streams showing but i do?

**Comment by Sainaamr** [comment]
now the bug is not there?

**Comment by Anishyou** [comment]
> now the bug is not there?

yes it is solved

File: lib/view_models/stream_view_model.dart [hunk 4]
```diff
@@ -25,52 +25,82 @@ class StreamViewModel extends StateNotifier<StreamState> {
...
+  }
+
+  /// Handles Fetching of a thumbnail for a live stream
+  Future<String> fetchStreamThumbnail(Int64 streamId) async {
     try {
-      final stream = await StreamHandler(_grpcHandler).fetchStream(streamId);
-      state = state.copyWith(streams: [stream], isLoading: false);
+      _logger.i('Fetching thumbnail VOD for stream ID: $streamId');
+      String thumbnailPath =
+          await StreamHandler(_grpcHandler).fetchThumbnailVOD(streamId);
```
Comment (GravityDarkLab) [inline lib/view_models/stream_view_model.dart:61]: should be like this?

```suggestion
          await StreamHandler(_grpcHandler).fetchThumbnailStreams(streamId);
```

File: lib/view_models/stream_view_model.dart
Comment (GravityDarkLab) [inline lib/view_models/stream_view_model.dart:1]: Looks good.
An Idea how you can refactor the thumbnails fetching:

```Dart
Future<void> fetchThumbnails() async {
    if (state.streams == null) {
      return;
    }
    var fetchThumbnailTasks = state.streams!
        .map((stream) => fetchThumbnailForStream(stream))
        .toList();

    var fetchedThumbnails = await Future.wait(fetchThumbnailTasks);
    state = state.copyWith(thumbnails: fetchedThumbnails, isLoading: false);
  }

  Future<String> fetchThumbnailForStream(Stream stream) async {
    try {
      return await (stream.liveNow
          ? fetchStreamThumbnail(stream.id)
          : fetchVODThumbnail(stream.id));
    } catch (e) {
      _logger.e('Error fetching thumbnail for stream ID ${stream.id}: $e');
      return '/thumb-fallback.png'; // Fallback thumbnail
    }
  }

  Future<String> fetchStreamThumbnail(Int64 streamId) async {
    try {
      _logger.i('Fetching thumbnail for live stream ID: $streamId');
      return await StreamHandler(_grpcHandler).fetchThumbnailStreams(streamId);
    } catch (e) {
      _logger.e('Error fetching thumbnail for live stream: $e');
      rethrow;
    }
  }

  Future<String> fetchVODThumbnail(Int64 streamId) async {
    try {
      _logger.i('Fetching thumbnail for VOD stream ID: $streamId');
      return await StreamHandler(_grpcHandler).fetchThumbnailVOD(streamId);
    } catch (e) {
      _logger.e('Error fetching thumbnail for VOD stream: $e');
      rethrow;
    }
  }
```

File: lib/views/course_view/components/course_card_view.dart [hunk 5]
```diff
@@ -1,5 +1,7 @@
 import 'package:flutter/material.dart';
 
+import '../../course_detail_view/course_detail_view.dart';
```
Comment (GravityDarkLab) [inline lib/views/course_view/components/course_card_view.dart:3]:
```suggestion
import 'package:gocast_mobile/views/course_detail_view/course_detail_view.dart';
```

File: lib/views/course_detail_view/course_detail_view.dart [hunk 6]
```diff
@@ -0,0 +1,102 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:gocast_mobile/providers.dart';
+import 'package:gocast_mobile/views/course_detail_view/stream_card.dart';
+import '../components/custom_search_top_nav_bar_back:button.dart';
```
Comment (GravityDarkLab) [inline lib/views/course_detail_view/course_detail_view.dart:5]:
```suggestion
import 'package:gocast_mobile/views/components/custom_search_top_nav_bar_back:button.dart';
```

**Review by GravityDarkLab — CHANGES_REQUESTED** [review]

**Review by GravityDarkLab — APPROVED** [review]
Looks good to me

**Review by Anishyou — APPROVED** [review]

**Review by cbetul — APPROVED** [review]
looks fine

**Pull request merged**